### ISSUE TITLE
all rt-hooks and all scripts share the same globals

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#3037]
Currently all scripts (which includes the run-time-hooks and the scripts listed on the command line) are executed in the same globals. The code in [`pyi_launch_run_scripts`](https://github.com/pyinstaller/pyinstaller/blob/v3.3/bootloader/src/pyi_launch.c#L363) can be emulated like this (I used this emulation to verify the bug and the solution):
```
import types

SCRIPTS = ['rt-hook-script.py', 'main-script.py']

MAIN = types.ModuleType('__main__')
MAIN_DICT = MAIN.__dict__

for filename in SCRIPTS:
    MAIN_DICT['__file__'] = filename
    with open(filename) as fh:
        src = fh.read()
    code = compile(src, filename, "exec")
    exec(code, MAIN_DICT, MAIN_DICT)
```
As you can see, the module is generated *once* and it's dict is used as scope for *all* scripts. This results into 1) later scripts having variables set from the earlier ones (which should not be a real problem given that "information leakage" is not an issue here) and 2) variables of earlier scripts may be overwritten by later ones. Actually the 2nd case showed up to be a problem (see #2949).

Thus the aforementioned code needs to be replace by something like:
```
for filename in SCRIPTS:
    MAIN = types.ModuleType('__main__')
    MAIN_DICT = MAIN.__dict__
    MAIN_DICT['__file__'] = filename
    with open(filename) as fh:
        src = fh.read()
    code = compile(src, filename, "exec")
    exec(code, MAIN_DICT, MAIN_DICT)
```

I already added a test-case for this: tests/functional/test_basic.py::test_several_scripts (the relevant code is at tests/functional/specs/several-scripts=.

**Comment by saneletm** [comment]
Does anyone know if Errors like this are related to this issue: 

3220 WARNING: Hidden import "wheezy.routing.router.PathRouter" not found!                                                                                     
53220 INFO: Loading module hook "hook-setuptools.py"...                                                                                                        
53666 INFO: Loading module hook "hook-PIL.SpiderImagePlugin.py"...                                                                                             
53672 INFO: Excluding import 'FixTk'                                                                                                                           
53674 INFO: Excluding import 'Tkinter'                                                                                                                         
53676 INFO: Loading module hook "hook-encodings.py"...                                                                                                         
54059 INFO: Loading module hook "hook-xml.etree.cElementTree.py"...                                                                                            
54060 INFO: Loading module hook "hook-numpy.core.py"...                                                                                                        
54108 INFO: Loading module hook "hook-sqlite3.py"...                                                                                                           
54139 INFO: Loading module hook "hook-django.py"...                                                                                                            
54909 INFO: Packages required by django:                                                                                                                       
['pytz']                                                                                                                                                       
54909 INFO: Django root directory /home/stm/DVR/emDjango1_11/evEnterpriseManager/evEnterpriseManager/evEnterpriseManager                                       
Loading avcodec version 58.35.100                                                                                                                              
Traceback (most recent call last):                                                                                                                             
  File "<string>", line 41, in <module>                                                                                                                        
  File "<string>", line 20, in walk_packages                                                                                                                   
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/post_office/tests/__init__.py", line 1, in <module>                                               
    from .test_backends import BackendTest                                                                                                                     
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/post_office/tests/test_backends.py", line 10, in <module>                                         
    from ..models import Email, STATUS, PRIORITY                                                                                                               
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/post_office/models.py", line 31, in <module>                                                      
    class Email(models.Model):                                                                                                                                 
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/db/models/base.py", line 110, in __new__                                                   
    app_config = apps.get_containing_app_config(module)                                                                                                        
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/apps/registry.py", line 247, in get_containing_app_config                                  
    self.check_apps_ready()                                                                                                                                    
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/apps/registry.py", line 125, in check_apps_ready                                           
    raise AppRegistryNotReady("Apps aren't loaded yet.")                                                                                                       
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.                                                                                            
Traceback (most recent call last):                                                                                                                             
  File "<string>", line 41, in <module>                                                                                                                        
  File "<string>", line 36, in walk_packages                                                                                                                   
  File "<string>", line 20, in walk_packages  
Traceback (most recent call last):                                                                                                                             
  File "<string>", line 41, in <module>                                                                                                                        
  File "<string>", line 20, in walk_packages                                                                                                                   
  File "/home/uname/DVR/emDjango1_11/evEnterpriseManager/evEnterpriseManager/failover/models/__init__.py", line 3, in <module>                                   
    from failover.models.core import *                                                                                                                         
  File "/home/uname/DVR/emDjango1_11/evEnterpriseManager/evEnterpriseManager/failover/models/core.py", line 5, in <module>                                       
    from ownership.models import CorruptFeatureOwnership                                                                                                       
  File "/home/uname/DVR/emDjango1_11/evEnterpriseManager/evEnterpriseManager/ownership/models.py", line 5, in <module>                                           
    class FeatureType(models.Model):                                                                                                                           
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/db/models/base.py", line 110, in __new__                                                   
    app_config = apps.get_containing_app_config(module)                                                                                                        
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/apps/registry.py", line 247, in get_containing_app_config                                  
    self.check_apps_ready()                                                                                                                                    
  File "/home/uname/Envs/django111/lib/python2.7/site-packages/django/apps/registry.py", line 125, in check_apps_ready                                           
    raise AppRegistryNotReady("Apps aren't loaded yet.")                                                                                                       
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.                                                                                            
Traceback (most recent call last):                                                                                                                             
  File "<string>", line 41, in <module>                                                                                                                        
  File "<string>", line 36, in walk_packages           

Basically looks like there is an issue with how pyinstaller after version 3.2 loads django models. 
I am running latest development version of pyisntaller with Python2.7.16 and Django1.11: 

The only pyinstaller change I have made is on PyInstaller/utils/hooks/subproc/django_import_finder.py, changing hiddenimports initialization to: 

hiddenimports = list(settings.INSTALLED_APPS) + \
     list(settings.TEMPLATES[0]['OPTIONS'].get('context_processors', [])) + \
     list(settings.TEMPLATES[0]['OPTIONS'].get('loaders', [])) + \
     [settings.ROOT_URLCONF]
